File: src/demos/python/ros/demo_ROS_custom_handler.py
# =============================================================================
# PROJECT CHRONO - http://projectchrono.org
#
# Copyright (c) 2026 projectchrono.org
# All rights reserved.
#
# Use of this source code is governed by a BSD-style license that can be found
# in the LICENSE file at the top level of the distribution and at
# http://projectchrono.org/license-chrono.txt.
#
# =============================================================================
# Authors: Aaron Young, Patrick Chen
# =============================================================================
#
# Demo: how to create a custom Chrono::ROS handler in Python.
#
# Subclass chros.ChROSHandler (a SWIG director) and address any installed ROS 2
# message type by its type-name string - no rclpy, no bridge code. This is the
# same handler as the Chrono 9.0 demo (an incrementing Int64 at 1 Hz), but it now
# runs through the schema bridge instead of in-process rclcpp.
#
#   ros2 topic echo /chrono_ros_node/my_topic
#
# =============================================================================

import pychrono as ch
import pychrono.ros as chros


class MyCustomHandler(chros.ChROSHandler):
    """A custom handler that publishes incrementing Int64 messages."""

    def __init__(self, topic):
        super().__init__(1)  # 1 Hz
        self.topic = topic
        self.publisher = None
        self.ticker = 0

    def Initialize(self, bridge):
        print(f"Creating publisher for topic {self.topic} ...")
        self.publisher = bridge.CreatePublisher(self.topic, "std_msgs/msg/Int64")
        return True

    def Tick(self, time):
        print(f"Publishing {self.ticker} ...")
        msg = self.publisher.NewMessage()
        msg["data"] = self.ticker
        self.publisher.Publish(msg)
        self.ticker += 1


def main():
    # Create the system.
    sys = ch.ChSystemNSC()
    sys.SetGravitationalAcceleration(ch.ChVector3d(0, 0, -9.81))

    phys_mat = ch.ChContactMaterialNSC()
    phys_mat.SetFriction(0.5)

    floor = ch.ChBodyEasyBox(10, 10, 1, 1000, True, True, phys_mat)
    floor.SetPos(ch.ChVector3d(0, 0, -1))
    floor.SetFixed(True)
    floor.SetName("base_link")
    sys.Add(floor)

    box = ch.ChBodyEasyBox(1, 1, 1, 1000, True, True, phys_mat)
    box.SetPos(ch.ChVector3d(0, 0, 5))
    box.SetRot(ch.QuatFromAngleAxis(0.2, ch.ChVector3d(1, 0, 0)))
    box.SetName("box")
    sys.Add(box)

    # Create the ROS manager and register the handlers.
    ros_manager = chros.ChROSManager()
    ros_manager.RegisterHandler(chros.ChROSClockHandler())
    ros_manager.RegisterHandler(chros.ChROSBodyHandler(25, box, "~/box"))

    tf_handler = chros.ChROSTFHandler(30)
    tf_handler.AddTransform(floor, floor.GetName(), box, box.GetName())
    ros_manager.RegisterHandler(tf_handler)

    ros_manager.RegisterHandler(MyCustomHandler("~/my_topic"))

    ros_manager.Initialize()

    # Simulation loop.
    time = 0
    time_step = 1e-3
    time_end = 30

    realtime_timer = ch.ChRealtimeStepTimer()
    while time < time_end:
        sys.DoStepDynamics(time_step)
        time = sys.GetChTime()

        if not ros_manager.Update(time, time_step):
            break

        realtime_timer.Spin(time_step)


if __name__ == "__main__":
    main()

File: src/demos/ros/demo_ROS_custom_handler.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2026 projectchrono.org
// All rights reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Authors: Aaron Young, Patrick Chen
// =============================================================================
//
// Demo: how to create a custom Chrono::ROS handler.
//
// Chrono::ROS is schema-driven: a handler addresses any installed ROS 2 message
// type by its type-name string and sets fields by name - no bridge code, no
// recompiling Chrono, no rclcpp. The custom handler below publishes an
// incrementing Int64 at 1 Hz on "~/my_topic".
//
//   ros2 topic echo /chrono_ros_node/my_topic
//
// =============================================================================

#include "chrono/core/ChTypes.h"
#include "chrono/core/ChRealtimeStep.h"
#include "chrono/physics/ChSystemNSC.h"
#include "chrono/physics/ChBody.h"
#include "chrono/physics/ChBodyEasy.h"

#include "chrono_ros/ChROSManager.h"
#include "chrono_ros/ChROSBridge.h"
#include "chrono_ros/ChROSHandler.h"
#include "chrono_ros/handlers/ChROSClockHandler.h"
#include "chrono_ros/handlers/ChROSBodyHandler.h"
#include "chrono_ros/handlers/ChROSTFHandler.h"

#include <iostream>

using namespace chrono;
using namespace chrono::ros;

// =============================================================================

class MyCustomHandler : public ChROSHandler {
  public:
    MyCustomHandler(const std::string& topic) : ChROSHandler(1), m_topic(topic), m_ticker(0) {}

    bool Initialize(ChROSBridge& bridge) override {
        std::cout << "Creating publisher for topic " << m_topic << " ..." << std::endl;
        m_publisher = bridge.CreatePublisher(m_topic, "std_msgs/msg/Int64");
        return true;
    }

    void Tick(double time) override {
        std::cout << "Publishing " << m_ticker << " ..." << std::endl;
        auto msg = m_publisher->NewMessage();
        msg.SetInt("data", m_ticker);
        m_publisher->Publish(msg);
        m_ticker++;
    }

  private:
    const std::string m_topic;
    std::shared_ptr<ChROSPublisher> m_publisher;
    int64_t m_ticker;
};

// =============================================================================

int main(int argc, char* argv[]) {
    std::cout << "Copyright (c) 2026 projectchrono.org\nChrono version: " << CHRONO_VERSION << std::endl << std::endl;

    // Create the system.
    ChSystemNSC sys;
    sys.SetGravitationalAcceleration({0, 0, -9.81});

    auto phys_mat = chrono_types::make_shared<ChContactMaterialNSC>();
    phys_mat->SetFriction(0.5f);

    auto floor = chrono_types::make_shared<ChBodyEasyBox>(20, 20, 1, 1000, true, true, phys_mat);
    floor->SetPos({0, 0, -1});
    floor->SetFixed(true);
    floor->SetName("floor");
    sys.AddBody(floor);

    auto box = chrono_types::make_shared<ChBodyEasyBox>(1, 1, 1, 1000, true, true, phys_mat);
    box->SetPos({0, 0, 5});
    box->SetRot(QuatFromAngleAxis(0.2, {1, 0, 0}));
    box->SetName("box");
    sys.AddBody(box);

    // ------------

    // Create the ROS manager and register the handlers.
    auto ros_manager = chrono_types::make_shared<ChROSManager>();

    // The clock automatically publishes on every tick on topic /clock.
    ros_manager->RegisterHandler(chrono_types::make_shared<ChROSClockHandler>());

    // Publish the state of the box at 25 Hz on the topic "~/box".
    ros_manager->RegisterHandler(chrono_types::make_shared<ChROSBodyHandler>(25, box, "~/box"));

    // Publish the transform of the box relative to the floor.
    auto tf_handler = chrono_types::make_shared<ChROSTFHandler>(100);
    tf_handler->AddTransform(floor, floor->GetName(), box, box->GetName());
    ros_manager->RegisterHandler(tf_handler);

    // The custom handler.
    ros_manager->RegisterHandler(chrono_types::make_shared<MyCustomHandler>("~/my_topic"));

    ros_manager->Initialize();

    // ------------

    // Simulation loop.
    double time = 0;
    const double step_size = 2e-3;
    const double time_end = 1000;

    ChRealtimeStepTimer realtime_timer;
    while (time < time_end) {
        time = sys.GetChTime();

        if (!ros_manager->Update(time, step_size))
            break;

        sys.DoStepDynamics(step_size);
        realtime_timer.Spin(step_size);
    }

    return 0;
}
